src/librustc_trans/abi.rs

```diff
@@ -244,8 +244,12 @@ impl Reg {
                     _ => bug!("unsupported float: {:?}", self)
                 }
             }
-            RegKind::Vector => {
-                Type::vector(&Type::i8(ccx), self.size.bytes())
+            RegKind::Vector { x86_mmx } => {
+                if x86_mmx {
```

Comment:
Any chance you could have the decision here, based solely on the size? Or is that not enough and you'd need the count as well?

Reply:
I do sort of forget why at this point but [this](https://github.com/alexcrichton/rust/blob/30b154dca4eff4550fac55bee7b53ebd4507c36a/src/librustc_trans/type_of.rs#L28-L36) was the main other location that actually made tests pass and this location of calling `Type::x86_mmx` was mostly for just getting everything to compile.

In that sense the linked location there already has the size/count and would avoid changing `RegKind::Vector`, but this location indeed only has the size as opposed to count as well. The current encoding of "a one-element vector 64-bits in size is x86_mmx" is a bit of a hack, so I could also change this to just look at the size here (unsure if there's other 64-bit vectors of interest on x86)

Do you know what this specific location is used for though in terms of the LLVM type returned? Maybe this doesn't need to get updated at all?

Reply:
Ahhh, you're right, this one is for call ABI casts, e.g. when `cabi_x86_64.rs` decides it wants to pass something in a SSE register instead of indirectly or in general-purpose registers.
As long as your *actual* argument/return `layout::Abi` is `Vector`, you shouldn't reach this path, nor should anything need MMX registers here (or at least I am not aware of any call ABI that uses them).